### ISSUE TITLE
Do not force upgrade container tag to be status.latestVersion

### ISSUE THREAD

**Issue by masterzen** [rancher/system-upgrade-controller#59]
This is somehow both a bug report and a feature request.

**Version**

v0.4.0

**Platform/Architecture**

Fedora CoreOS

**Describe the bug**

We were planning to use the System Upgrade Controller to upgrade a fleet of Fedora CoreOS VM. The way the System Upgrade Controller is working fits our model: we could update the plan `version` field with the latest released Fedora CoreOS version (manually or automatically), and the upgrade container would just be an `alpine` container that `chroot` in the image and run the necessary commands to upgrade the OS and reboot to the intended `version`.

Except that it doesn't work because the System Upgrade Controller substitutes the `upgrade` image tag (provided in the plan) by the `status.latestVersion` (ie the plan `version`).
I understand that's how rancher is updated, but that breaks a whole lot of workflows where the upgrade container image is not that important.

I understand that I can workaround this and publish locally an alpine image tagged with the current version (possibly in the `prepare` container, since it's not yet subject to tag overwriting see #41 ), but that'd be awesome, IMHO to not overwrite a specified image tag in both the prepare or upgrade container:

 IMHO, the behavior should be:
* if the `upgrade` or `prepare` container image has the form `image:tag`, then use this explicitely
* if the `upgrade` or `prepare` container image has the form `image` (without a tag) then use the `latestVersion` as the image tag.

This allows both workflows to coexist.

**To Reproduce**

Use the following plan:

```yaml
apiVersion: upgrade.cattle.io/v1
kind: Plan

metadata:
  name: fedora-coreos-asmodee-production
  namespace: system-upgrade
  labels:
    addonmanager.kubernetes.io/mode: Reconcile

spec:
  concurrency: 1
  version: "31.20200323.3.2"
  nodeSelector:
    matchExpressions:
      - {key: role, operator: In, values: ["whatever","dont-care"]}

  serviceAccountName: system-upgrade

  secrets:
    - name: fedora-coreos
      path: /host/run/system-upgrade/fedora-coreos
  upgrade:
    image: alpine:3.11
    command: ["chroot", "/host"]
    args: ["/usr/bin/bash", "/run/system-upgrade/fedora-coreos/upgrade.sh"]
```

**Expected behavior**
It should run the `/run/system-upgrade/fedora-coreos/upgrade.sh` on every node after `chroot`ing to the host.

**Actual behavior**

It fails scheduling the upgrade pod on the nodes with:
```
Failed to pull image "alpine:31.20200323.3.2": rpc error: code = Unknown desc = Error response from daemon: manifest for alpine:31.20200323.3.2 not found
```

**Comment by dweomer** [comment]
Hi @masterzen,

This isn't perfect as it somewhat obfuscates the input version but you can use an upgrade container with a tag/version that doesn't change combined with a secret. The data from **Secret**s referenced by **Plan**s are calculated along with the `.status.latestVersion` (and `.spec.serviceAccountName`) which is what SUC tags nodes with to know to not attempt further applications of a **Plan**.

Below is an example of such that I demo'd today at the CNCF webinar:
```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: kernel
  namespace: system-upgrade
type: Opaque
stringData:
  version: 5.3.0.46.102
  upgrade.sh: |
    #!/bin/sh
    set -e
    export DEBIAN_FRONTEND=noninteractive
    secrets=$(dirname $0)
    apt-get update
    apt-get -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install -yq \
      linux-virtual-hwe-18.04=$(cat $secrets/version) \
      linux-headers-virtual-hwe-18.04=$(cat $secrets/version) \
      linux-image-virtual-hwe-18.04=$(cat $secrets/version)
    if [ -f /run/reboot-required ]; then
      cat /run/reboot-required
      reboot
    fi
---
apiVersion: upgrade.cattle.io/v1
kind: Plan
metadata:
  name: kernel
  namespace: system-upgrade
spec:
  # The maximum number of concurrent nodes to apply this update on.
  concurrency: 2

  # Select which nodes this plan can be applied to.
  nodeSelector:
    matchExpressions:
      - {key: kernel-upgrade, operator: Exists}
      - {key: kernel-upgrade, operator: NotIn, values: ["disabled", "false"]}

  # The service account for the pod to use. As with normal pods, if not specified the `default` service account from the namespace will be assigned.
  # See https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
  serviceAccountName: system-upgrade

  # Link to secrets that will be mounted into all containers in job pods.
  # Secrets are calculated as part of the `.status.latestHash`. This means that changing
  # the secrets profile for a plan or changing the secrets that a plan refers to will
  # trigger application.
  secrets:
    - name: kernel
      # optional: default=/run/system-upgrade/secrets/{name}
      path: /host/run/system-upgrade/secrets/kernel

  # The value for `channel` is assumed to be a URL that returns HTTP 302 with the last path element of the value
  # returned in the Location header assumed to be an image tag.
  #channel: https://canonical.example.com/ubuntu/lts

  # Providing a value for `version` will prevent polling/resolution of the `channel` if specified.  
  version: bionic

  # The prepare init container is run before cordon/drain which is run before the upgrade container.
  # Shares the same format as the `upgrade` container
  #prepare:
  #  image: alpine:3.11
  #  command: [sh, -c]
  #  args: [" echo '### ENV ###'; env | sort; echo '### RUN ###'; find /run/system-upgrade | sort"]

  # See https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service
  drain:
    # deleteLocalData: true
    # ignoreDaemonSets: true
    force: true

  # A very limited container spec
  upgrade:
    # The tag portion of the image will be overridden with the value from `.status.latestVersion` a.k.a. the resolved version.
    image: ubuntu
    command: ["chroot", "/host"]
    args: ["sh", "/run/system-upgrade/secrets/kernel/upgrade.sh"]
```

So, instead of upgrading your **Plan**'s `.spec.version` you can upgrade what is essentially an underlying configuration parameter (the keys/values of the **Secret's** underlying map) and that will trigger (re-)application of your plan.

**Comment by masterzen** [comment]
Hi @dweomer ,

Thanks for the workaround. Indeed this would solve my current issue. Still that doesn't allow to use the Plan's channel for automatic upgrade discovery (which is what I ultimately plan to do with our Fedora CoreOS systems, by creating a proxy between Cincinnati and the System Upgrade Controller channel mechanism).

To fix our immediate issue, I've implemented my proposed change in PR #60. Feel free to review and discuss it. If needed the behavior can be controlled by an executable flag or option to prevent the backward incompatibility.

**Comment by dweomer** [comment]
@masterzen as mentioned in #61, reviewing #60 clued me into some broken code+tests. I've merged #61 and cut a milestone release [v0.5.0-m1](https://github.com/rancher/system-upgrade-controller/releases/tag/v0.5.0-m1). Can you try that out and let me know if it works for your use-case?

**Comment by masterzen** [comment]
@dweomer,

Many thanks for your work !

I was able to test v0.5.0-m1 and it worked as intended in our upgrade workflow (ie image tag hasn't been overwritten). I read the PR and didn't see any issue.

Again, many thanks for the quick fix !
